### PR TITLE
Raise exception when HasProps receives invalid kwargs

### LINKED ISSUE #1960: Throw exception if nonexistent keyword arguments are given
There was a typo in the line `glyphs/line.py` example where it said `line_with` instead of `line_width`. Not a big deal in this case, but can lead to problems that are hard to debug. My feeling is that we should raise an error, or at least show a warning when an invalid keyword argument is given.

### PR DESCRIPTION
**Pull request by almarklein**
issues: closes #1960

In `_glyph_function()` we pop all `kwargs` that we need and pass the rest to the `Glyph` class (and effectively to `HasProps.__init__`).

### CODE REVIEW

**Comment by almarklein** [comment]
WIP, but could use some feedback whether this seems a sensible approach. Also, this would need extensive testing, because using a non-existent property in *any* HasProps class will now raise an error.

File: bokeh/properties.py [hunk 1]
```diff
@@ -591,6 +591,9 @@ def __init__(self, **properties):
         self._changed_vars = set()
 
         for name, value in properties.items():
+            if not hasattr(self, name):
+                raise ValueError('Class %r does not have property %r' % 
+                                 (self.__class__.__name__, name))
```
Comment (mattpap) [inline bokeh/properties.py]: This is very unnecessary. See a couple lines below.

Comment (almarklein) [inline bokeh/properties.py]: Ah, I overlooked that. Thanks.

File: bokeh/_glyph_functions.py [hunk 2]
```diff
@@ -52,14 +59,19 @@ def func(plot, *args, **kwargs):
...
-        glyph_props = glyphclass.properties() | set(argnames)
-        glyph_kwargs = dict((key, value) for (key, value) in iteritems(kwargs) if key in glyph_props)
-        glyph = glyphclass(**glyph_kwargs)
-
+    
         nonselection_glyph_params = _materialize_colors_and_alpha(kwargs, prefix='nonselection_', default_alpha=0.1)
+        
+        #glyph_props = glyphclass.properties() | set(argnames)
+        #glyph_kwargs = dict((key, value) for (key, value) in iteritems(kwargs) if key in glyph_props)
+        #glyph = glyphclass(**glyph_kwargs)
```
Comment (bryevdv) [inline bokeh/_glyph_functions.py]: prefer to not check in commented out code

**Comment by bryevdv** [comment]
`HasProps` classes already raise an error on unknown properties:
```
In [5]: from bokeh.models import Range1d

In [6]: Range1d(foo=10)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-d6ba390ca24b> in <module>()
----> 1 Range1d(foo=10)

/Users/bryan/work/bokeh/bokeh/models/ranges.py in __init__(self, *args, **kwargs)
     44             kwargs['start'] = args[0]
     45             kwargs['end'] = args[1]
---> 46         super(Range1d, self).__init__(**kwargs)
     47 
     48 

/Users/bryan/work/bokeh/bokeh/plot_object.py in __init__(self, **kwargs)
     88 
     89         if not block_events:
---> 90             super(PlotObject, self).__init__(**kwargs)
     91             self.setup_events()
     92         else:

/Users/bryan/work/bokeh/bokeh/properties.py in __init__(self, **properties)
    592 
    593         for name, value in properties.items():
--> 594             setattr(self, name, value)
    595 
    596     def __setattr__(self, name, value):

/Users/bryan/work/bokeh/bokeh/properties.py in __setattr__(self, name, value)
    606 
    607             raise AttributeError("unexpected attribute '%s' to %s, %s attributes are %s" %
--> 608                 (name, self.__class__.__name__, text, nice_join(matches)))
    609 
    610     def clone(self):

AttributeError: unexpected attribute 'foo' to Range1d, possible attributes are end, name, session, start or tags

In [7]: r = Range1d(10, 20)

In [8]: r.foo = 10
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-8-b8a1b0b61dd1> in <module>()
----> 1 r.foo = 10

/Users/bryan/work/bokeh/bokeh/properties.py in __setattr__(self, name, value)
    606 
    607             raise AttributeError("unexpected attribute '%s' to %s, %s attributes are %s" %
--> 608                 (name, self.__class__.__name__, text, nice_join(matches)))
    609 
    610     def clone(self):

AttributeError: unexpected attribute 'foo' to Range1d, possible attributes are end, name, session, start or tags
```
But some classes/functions accept additional arguments as inputs, and then should pass the "rest" on to a `HasProps` class. Perhaps there are places where some logic problems are throwing unrecognized args on the floor without warning. 

I think this pattern (props arg + non-props args) has become common enough, that we should attempt to codify it some. That is, functions that accept `kwargs` that are a mix of things that are plain arguments plus some that get passed to `HasProps`, should have to specify the "non-props` args somehow. This will allow to strip them out automatically and also help with validation.

**Comment by bryevdv** [comment]
The case here is also a tricky one, the `kwargs` are a mix of three things:
* plain parameters for the function ('alpha', 'color')
* parameters for the glyph renderer ('x_range_name', etc)
* parameters for a glyph
I'm not sure what the best way to improve spelling/validation in this particular case is. @mattpap, thoughts?

File: bokeh/_glyph_functions.py (hunk 2 above)
Comment (almarklein) [inline bokeh/_glyph_functions.py]: I agree. This is still WIP and was waiting for your blessing with this approach before really going forward with this :)

**Comment by almarklein** [comment]
Yeah, I overlooked that HasProps already does the checking. This is a comforting thought :)  

>  functions that accept kwargs that are a mix of things that are plain arguments plus some that get passed to HasProps, should have to specify the "non-props" args somehow.

I actually started out with something along these lines, but I did not like the fact that one has to specify things twice. DRY and all that. 

IMO we should be consistent about *popping* the non-props from the `kwargs` and pass the rest to `HasProps`. In this particular case it's a bit more complicated due to the color and alpha args, but the args for the renderer are easily popped (the popping of the renderer args can probably be grouped better to make the code more clear though).

**Comment by bryevdv** [comment]
Well, what I was proposing was not a DRY violation, per se. I was proposing:
* a spec for "non-prop" args (which does not exist anywhere currently, so not a repetion),
* those values get automatically popped off kwargs
* the rest get passed to a `HasProps` where validation finds and errors on anything extra

As mentioned, that is not quite sufficient for this particular case, where there are two `HasProps` objects to create. 

One solution would be to have a method on `HasProps` class that will "pre-extract" a dict with just the args that it might care about. Then those could be passed to the initializer. In cases like this it would look like:
```
    non_props = process(kwargs)  # pops off values

    a_props = A.check(kwargs) # pops off values
    a_obj = A(a_props)

    b_props = B.check(kwargs) # pops off values
    b_obj = B(b_props)

   # kwargs should be empty at this point
   
    # alternatively: b_obj = B(kwargs) to consume last kwargs 
```

**Comment by almarklein** [comment]
> a spec for "non-prop" args

How do you see this? Maybe something like this?

    subset = pop_kwargs(kwargs, 'names', 'of_args', 'that_we_need')   # where subset if a dict

In this case, the kwargs intended for the renderer are known, and I feel safer extracting them explicitly; if renderer has a property that Glyph has too, we want to reserve it for the glyph ...

**Comment by bryevdv** [comment]
Well there is tradeoffs. The problem with asking and then extracting and then passing in only the "right" ones is that you lose the benefit that started this whole thread: having "extra" parameters trigger errors. Unless you remember to check for an empty dict every single time, but that is an annoying and error prone burden. The common case is "func has some args, and (one) hasprops obj has the rest" so I would optimize for that:

```
@args_thing("foo", "bar" ,baz")
def myfunc(func_kwargs, the_rest):
    
    # func_kwargs dict has "foo", "bar", and "baz"
 
    obj = SomeModel(the_rest)  # any extras throw an error

myfunc(foo=10, bar=20, baz=30, a=1, b=2)

```

**Comment by bryevdv** [comment]
This could certainly be made even more sophisticated to accommodate required args. We also incorporate the "asking and extracting". This I believe is the most general case to deal with:
```
@args_thing(pos=["foo", "bar"], kw=["baz", "quux"], models=[ModelA, ModelB], doc="""

One hard-ish part is making this play nicely with autodoc systems. The signature *should* 
show up as:

    myfunc(foo, bar, **kw)

in the documentation. We would might have to make another sphinx plugin to handle that.

""")
def myfunc(foo, bar, func_kwargs, a_kwargs, the_rest):
    
     # func_kwargs  has "baz" and "quux"
    
     # a_kwargs has args for ModelA
    a = ModelA(**a_kwargs)

    # let ModelB check the rest
    b = ModelB(**the_rest)
```

**Comment by almarklein** [comment]
That is certainly a possibility, but seems a bit overkill IMO. What about this:

    # Most functions can just do this
    def foo(spam=2, eggs=3, **rest):
         """ blabla, we use spam and eggs and use the rest to create model X.
         """
        ...
         a = ModelA(**rest)

     # Alternative (we already do this in may places)
     def foo(**kwargs):
          spam = kwargs.pop('spam', 2)
          eggs = kwargs.pop('eggs', 4)
          a = ModelA(**kwargs)

    # And for easily extracting a larger subset
    def  foo(**kwargs):
        model_a_kwargs = pop_items(kwargs, bar, spam=2, eggs=4)
        # this new dict will always have spam and eggs, 
        # and have bar if it was present in kwargs
        a = ModelA(**model_a_kwargs)
        b = ModelB(**kwargs)  # use the rest
   
That `pop_items` function is just 7 lines of code, and is basically a convenience function to pop multiple args at once.

**Comment by almarklein** [comment]
`pop_items` can be simpler. Just accept string args (edited previous comment)

**Comment by bryevdv** [comment]
Well the reason I am suggesting it is because it provides a *specification*, and specifications can be automatically documented correctly. If there is just `foo(**kw)` then the only thing that says what the actual params are is the code, and if someone hand-writes one, the docstring. If the docstring is handwritten, it will fall out of date at some point, guaranteed. 

Basically there has been no instance where increased automation of testing and documentation has not been a really good thing, so I am now much less resistant to putting a little effort into "infrastructure".  Especially if once it is done it means we can forget about having to worry whether the docs go out of date, ever. Or if it means we have one common pattern that can itself be documented in dev docs for potential new devs to understand. (Or in this case, both)

**Comment by bryevdv** [comment]
Basically I am now of the mind than anything and everything that can be automated, absolutely *should* be automated. That is because there is already more to manage and maintain than we can keep up with. The *only* thing that **reduces** our overall burden in any appreciable way, is automation.

**Comment by bryevdv** [comment]
Also to clairfy, `pop_items` is an improvement, so there is no reason not to do this right now. But I would like to explore codify specifications for these interfaces more rigidly at some point.

**Comment by almarklein** [comment]
I understand that, but in this particular case it seems to make less sense since we're talking about a function that is not public (the function `func` inside `_glyph_function`). I suppose this is why I'm now biased towards a simpler solution. 

This PR currently addresses the dropping of keyword arguments in this particular function. I do not know whether this occurs in other places, or whether defining multiple models from a single `kwargs` dict occurs in many other places. If this is a rare situation, I'd rather fix it in a simple (and quick) way, instead of spending time on trying to automate it (writing and maintaining that sphinx plugin is a burden too).

File: bokeh/_glyph_functions.py [hunk 3]
```diff
@@ -22,20 +23,14 @@ def func(plot, *args, **kwargs):
...
-        legend_name = kwargs.pop("legend", None)
-
+        
         if not isinstance(plot, Plot):
             raise ValueError("expected plot object for first argument")
-
+        
         # TODO (bev) this seems like it should be here but invalid kwargs
         # currently get through (see also below)
         # plot.update(**kwargs)
```
Comment (bryevdv) [inline bokeh/_glyph_functions.py]: These three lines can be removed entirely. They date back to when there were "glyph functions" that also produced plots themselves.

**Comment by bryevdv** [comment]
Well I do know that this and similar patterns do occur in multiple places, and have been the cause of bugs and problems. I probably should have been clear that it was experience motivating these suggestions. As I said `popitem` is an improvement too, we should proceed with here and automation in a different PR at some point.

**Comment by almarklein** [comment]
Ok, in that case we should dive deeper at some point.

**Comment by almarklein** [comment]
Sweet! Travis reports tons of invalid properties... Will do some fixing tomorrow...

**Comment by bryevdv** [comment]
OK, this will require a little more. The ones like:
```
AttributeError: unexpected attribute 'outer_radius_units' to AnnularWedge, similar attributes are outer_radius or inner_radius
[FAIL] /home/travis/build/bokeh/bokeh/examples/plotting/file/glyphs.py
```
should actually pass through. Adding "_units" to an existing data spec property is the convenience way to specify its units, but it translates into an internal configuration of the data spec property, not a separate property of its own. 

All the other ones do look like they should be fixed up (`toolbar_location` to glyph funcs, etc. goes back to some original API ideas)

**Comment by almarklein** [comment]
Ok, thanks for the heads-up on the unit thing!

**Comment by damianavila** [comment]
>anything and everything that can be automated, absolutely should be automated

I will write this sentence in a t-shirt :stuck_out_tongue_winking_eye:

**Comment by almarklein** [comment]
> anything and everything that can be automated, absolutely should be automated

Not *always* imo: http://xkcd.com/1319/

**Comment by almarklein** [comment]
Arg! There are occasions (e.g. here https://github.com/bokeh/bokeh/blob/master/bokeh/transforms/ar_downsample.py#L452) where we just push a bunch of plot_options to the glyph function. Some options:

* Forget about it; we just allow throwing anything in kwargs, and if there was a typo its tough luck. This seems a bit harsh, but allows the code to be less strict/complex.
* Maybe we can somehow distinguish code-generated properties from user-defined properties. The former can then be ignored. Not a perfect solution, because code is written by humans too.
* Any code that subits kwargs to a function, must make sure that that function can handle all kwargs.

**Comment by almarklein** [comment]
The last one is hard, e.g. `_glyph_function()/func()` will use the kwargs in different ways ...  Maybe with that spec that @bryevdv mentioned, a function could expose what kwargs it accepts ...

**Comment by damianavila** [comment]
>Not always imo: http://xkcd.com/1319/

If you go to the second "window" (the reality), then you did not know well your problem. If I try to automate something is because I am sure I will "reproduce" the theory... in any case, a discussion to have with some :beers: in the middle...

>The last one is hard, e.g. _glyph_function()/func() will use the kwargs in different ways ... Maybe with that spec that @bryevdv mentioned, a function could expose what kwargs it accepts ...

I agree it is hard, but it would be interesting and also, probably, outside the scope of this PR... so from the three options, I guess the first one is the path to follow for now...

**Comment by bryevdv** [comment]
> I agree it is hard, but it would be interesting and also, probably, outside the scope of this PR... so from the three options, I guess the first one is the path to follow for now...

Why? This has been a problem and something people have been trying to fix in various ways for over a year now. Let's make a plan and fix it, for real.

**Comment by damianavila** [comment]
>Why? This has been a problem and something people have been trying to fix in various ways for over a year now. Let's make a plan and fix it, for real.

I don't mean not to fix it... I suggest make the plan in an **Issue** and and fix it in another PR... but if you want to discuss it here... no problem. 

In general, I am just trying to get the different aspect of the discussion a little but separated and trying to encouraged people to have the discussion about design things in Issues (not PRs). The PRs discussion have to be related to the implementation itself and I know this PR has pertinent implementation discussions, but it  also has design discussion... and getting those discussion exposed in Issues help us not only to us, but also to all the new contributors who wants to know more about the Bokeh internal... does make sense? (for me yes :wink:)

**Comment by bryevdv** [comment]
OK I am fine with moving the discussion to the issue, but in that case maybe we should retire this PR.

**Comment by damianavila** [comment]
Closing accordingly to BEP-1, please reopen it when it is ready again. Thanks

**Pull request closed without merging**